src/gui/editors/PianoRoll.cpp

```diff
@@ -2839,165 +2752,146 @@ void PianoRoll::dragNotes(int x, int y, bool alt, bool shift, bool ctrl)
+			TimePos minLength = !alt ? quantization() : 1;
+			for (Note* note: notes)
+			{
+				if (note->oldPos() < selStart)
+				{
+					TimePos minEndPos = note->oldPos() + std::min(minLength, note->oldLength());
+					if (minEndPos >= selStart)
+					{
+						noteOffset = std::max(0, noteOffset);
+						break;
+					}
+					noteOffset = std::max(minEndPos - selEnd, noteOffset);
+				}
+			}
 		}
 
 		// Apply offset to all selected notes
 		for (Note *note : getSelectedNotes())
 		{
 			// Quick resize is only enabled on Nudge mode, since resizing the note
 			// while in Snap mode breaks the calculation of the note offset
-			if (shift && ! m_startedWithShift && m_gridMode == gridNudge)
+			if (shift && !m_startedWithShift && m_gridMode == gridNudge && m_editMode == ModeDraw)
 			{
 				// quick resize, toggled by holding shift after starting a note move, but not before
-				int ticks_new = note->oldLength().getTicks() + noteOffset;
-				if( ticks_new <= 0 )
-				{
-					ticks_new = 1;
-				}
-				note->setLength( TimePos( ticks_new ) );
-				m_lenOfNewNotes = note->length();
+				note->setLength(std::max(1, note->oldLength() + noteOffset));
+				if (note == m_currentNote) { m_lenOfNewNotes = note->length(); }
 			}
 			else
 			{
 				// moving note
-
-				// Final position of the note
-				TimePos posTicks(note->oldPos().getTicks() + noteOffset);
-				int key_num = note->oldKey() + off_key;
-
-				note->setPos(posTicks);
-				note->setKey(key_num);
+				note->setPos(note->oldPos() + noteOffset);
+				note->setKey(note->oldKey() + off_key);
+			}
+		}
+		// Resize notes left of the selection in bulldozer mode
+		if (m_action == ActionBulldozerMove)
+		{
+			for (Note* note: notes)
+			{
+				if (note->selected()) { continue; }
+				// Move end-points that touched or overlaped the selection initially
+				else if (note->oldEndPos() >= selStart)
+				{
+					note->setLength(note->oldLength() + noteOffset);
+				}
+				// Squash notes that overlaps the selection now
+				else if (note->oldEndPos() > selStart + noteOffset)
+				{
+					note->setLength(selStart + noteOffset - note->oldPos());
+				}
 			}
 		}
+
 	}
-	else if (m_action == ActionResizeNote)
+	else if (m_action & ActionResizeAny)
 	{
-		// When resizing notes:
-		// If shift is not pressed, resize the selected notes but do not rearrange them
-		// If shift is pressed we resize and rearrange only the selected notes
-		// If shift + ctrl then we also rearrange all posterior notes (sticky)
-		// If shift is pressed but only one note is selected, apply sticky
-
 		// Quantize the resizing if alt is not pressed
-		if (!alt)
+		if (!alt && quantization() > 1)
 		{
-			off_ticks = floor(off_ticks / quantization()) * quantization();
+			if (m_gridMode == gridSnap)
+			{
+				TimePos end = m_currentNote->oldEndPos();
+				off_ticks = Note::quantized(end + off_ticks, quantization()) - end;
+			}
+			else
+			{
+				off_ticks = floor(off_ticks / quantization()) * quantization();
```

Comment:
Didn't know about `::min()` and `::max()`, but I'll leave it as is for now, since the same line appears a few lines up.

FTR I think we should change `TimePos::quantize` (why does it need a float?) `Note::quantized` (breaks for negative values + why inside Note??) and this calculation (should *probably* round instead of floor), but it goes out of scope for this PR.

Reply:
> FTR I think we should change TimePos::quantize (why does it need a float?) Note::quantized (breaks for negative values + why inside Note??) and this calculation (should probably round instead of floor), but it goes out of scope for this PR.

Yeah, I feel like there are so many QoL changes that can be made to improve the codebase. Even myself when I look at the codebase it makes me question "why?" :laughing:.

However, I did read that you were eyeing a certain ``PianoRoll`` rewrite. Is there any PR/branch for this or is this just something you have in mind?

Reply:
Veratil has started 6063